components/camel-graphql/src/main/java/org/apache/camel/component/graphql/GraphqlEndpoint.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.component.graphql;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Map;
+
+import org.apache.camel.Component;
+import org.apache.camel.Consumer;
+import org.apache.camel.Processor;
+import org.apache.camel.Producer;
+import org.apache.camel.RuntimeCamelException;
+import org.apache.camel.spi.Metadata;
+import org.apache.camel.spi.UriEndpoint;
+import org.apache.camel.spi.UriParam;
+import org.apache.camel.spi.UriPath;
+import org.apache.camel.support.DefaultEndpoint;
+import org.apache.camel.util.IOHelper;
+import org.apache.camel.util.ObjectHelper;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+
+@UriEndpoint(firstVersion = "3.0.0", scheme = "graphql", title = "GraphQL", syntax = "graphql:httpUri", label = "graphql")
+public class GraphqlEndpoint extends DefaultEndpoint {
+
+    private final CloseableHttpClient httpclient;
+    @UriPath
+    @Metadata(required = true)
+    private URI httpUri;
+    @UriParam
+    private String query;
+    @UriParam
+    private String queryFile;
+    @UriParam
+    private String operationName;
+    @UriParam(prefix = "variable.", multiValue = true)
+    private Map<String, Object> variables;
```

Comment:
@oscerd Camel is failing to map the variable.* URI options to this class field. What's wrong?

Reply:
I get the following error:
```
Caused by: org.apache.camel.ResolveEndpointFailedException: Failed to resolve endpoint: graphql://http://localhost:43663/graphql?operationName=BookById&queryFile=multipleQueries.graphql&variable.id=book-1 due to: There are 1 parameters that couldn't be set on the endpoint. Check the uri if the parameters are spelt correctly and that they are properties of the endpoint. Unknown parameters=[{variable.id=book-1}]
```

Reply:
@alexkieling AFAIK the way the binding works in camel is by generating a property configurter java classes to see these properties and avoid reflections. However of course, it needs to case the input into the suitable type using [type converters](https://camel.apache.org/manual/latest/type-converter.html), meaning when it tries to set variables, first Camel will try to convert it from Object to Map. However in this case, I don't think Camel will be able to convert it out of box for you since there is a type converter from object to Map that shipped in (as I know but not 100%). In order to solve this, either you handle setting this by yourself (see here for an [example](https://github.com/apache/camel/blob/master/components/camel-http/src/main/java/org/apache/camel/component/http/HttpComponent.java#L321) on how camel-http handles that). Or you can add a type converter to convert from `Object` to `Map` and let Camel does it for you (an [example](https://github.com/apache/camel/blob/master/components/camel-debezium/src/main/java/org/apache/camel/component/debezium/DebeziumTypeConverter.java))

Reply:
@omarsmak Thanks for your answer. An alternative way would be defining my `variables` field as String and accept a comma-separated list of colon-separated key/value pairs, like `variables=var1:foo,var2:bar`. That strategy is used in the Kafka component. What do you think?
https://github.com/apache/camel/blob/35184b640382923fc52b56d38dda6f329ec23f9a/components/camel-kafka/src/test/java/org/apache/camel/component/kafka/KafkaComponentTest.java#L42

Reply:
Sure whatever you think is more convenient for the user and easy for you to work with, just bear in mind that you will still need to parse that by yourself in case you want to create the `Map` out of it.